src/main/java/com/palantir/gradle/versions/VirtualPlatformSettingsPlugin.java

```diff
@@ -0,0 +1,142 @@
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.immutables.value.Value;
+
+public class VirtualPlatformSettingsPlugin implements Plugin<Settings> {
+
+    private static final Logger log = Logging.getLogger(VirtualPlatformSettingsPlugin.class);
+
+    @Override
+    public final void apply(Settings settings) {
+        settings.getGradle().allprojects(project -> {
+            project.getDependencies().getComponents().all(VirtualPlatformRule.class);
+            project.getBuildscript().getDependencies().getComponents().all(VirtualPlatformRule.class);
+        });
+    }
+
+    public abstract static class VirtualPlatformRule implements ComponentMetadataRule {
+        private static final ObjectMapper XML_MAPPER =
+                new XmlMapper(new WstxInputFactory(), new WstxOutputFactory()).registerModule(new Jdk8Module());
+
+        @Inject
+        protected abstract RepositoryResourceAccessor getRepositoryResourceAccessor();
+
+        @Override
+        public final void execute(ComponentMetadataContext context) {
+            ModuleVersionIdentifier id = context.getDetails().getId();
+
+            String pomPath = buildPomPath(id);
+            getRepositoryResourceAccessor().withResource(pomPath, resource -> {
+                parseMetadata(resource)
+                        .flatMap(Metadata::extractDependencies)
+                        .filter(deps -> hasVirtualPlatform(deps, id.getGroup()))
+                        .ifPresent(_deps -> assignToPlatform(context, id));
+            });
+        }
+
+        private static Optional<Metadata> parseMetadata(InputStream resource) {
+            try {
+                return Optional.of(XML_MAPPER.readValue(resource, Metadata.class));
+            } catch (IOException e) {
+                log.debug("Failed to parse POM metadata: {}", e.getMessage());
+                return Optional.empty();
+            }
+        }
+
+        private static boolean hasVirtualPlatform(List<Dependency> dependencies, String expectedGroup) {
+            return dependencies.stream().anyMatch(dep -> isVirtualPlatformDependency(dep, expectedGroup));
+        }
+
+        private static boolean isVirtualPlatformDependency(Dependency dependency, String expectedGroup) {
+            return dependency.group().filter(expectedGroup::equals).isPresent()
+                    && dependency
+                            .module()
+                            .filter(AddVirtualPlatformPlugin.VIRTUAL_PLATFORM_NAME::equals)
+                            .isPresent();
+        }
+
+        private static String buildPomPath(ModuleVersionIdentifier id) {
+            String groupPath = id.getGroup().replace('.', '/');
+            return String.format(
+                    "%s/%s/%s/%s-%s.pom", groupPath, id.getName(), id.getVersion(), id.getName(), id.getVersion());
+        }
```

Comment:
If we could use `project.getDependencies().createArtifactResolutionQuery` this might be a bit nicer in that you wouldn't need to build up this path manually. `DependencyHandler` is listed as `@ServiceScope({Scope.Project.class})` so you _might_ be able to inject it here, there is a lot of mutable stuff on it though so it might be angry - worth a try?